serio/src/future.rs

```diff
@@ -0,0 +1,10 @@
+use std::future::Future;
+
+// Just a helper function to ensure the futures we're returning all have the
+// right implementations.
+pub(crate) fn assert_future<T, F>(future: F) -> F
```

Comment:
Do you think this has any performance impact, because now every future is copied into this function. I could imagine that this is optimized by the compiler.